### PR TITLE
Create external WeightedMath contract

### LINKED ISSUE #1919: Create `ExternalWeightedMath` contract
Wrap `WeightedMath` / `Math` in a contract that can be used as a external library to reduce bytecode size for managed pools.

### PR DESCRIPTION
**Pull request by jubeira**
The new `ExternalWeightedMath` wraps `WeightedMath`, `WeightedJoinsLib` and `WeightedExitsLib` so that they can be used (optionally) as an external library, reducing bytecode size.

`ManagedPool` and its base factory were added a new argument to support this external version of weighted math, and now it performs external calls instead of using the libraries directly.

### Results
`hh compile && cat artifacts/contracts/managed/ManagedPool.sol/ManagedPool.json | jq -r '.deployedBytecode' | wc --chars | awk '{print (($1-2)/2)/1024}'`

- master: 29,7036
- this change @ https://github.com/balancer-labs/balancer-v2-monorepo/pull/1921/commits/9e461a1eeadd327f19ada9703ce57a70f65278f1: 27,7603

Closes #1919.

### CODE REVIEW

**Comment by jubeira** [comment]
Test failing is flaky; solution should be #1923

File: pkg/pool-weighted/contracts/managed/BaseManagedPoolFactory.sol [hunk 1]
```diff
@@ -34,10 +34,14 @@ import "./ManagedPool.sol";
  * to deploy the pool, passing in that contract address as the owner.
  */
 contract BaseManagedPoolFactory is BasePoolFactory, FactoryWidePauseWindow {
-    constructor(IVault vault, IProtocolFeePercentagesProvider protocolFeeProvider)
-        BasePoolFactory(vault, protocolFeeProvider, type(ManagedPool).creationCode)
-    {
-        // solhint-disable-previous-line no-empty-blocks
+    ExternalWeightedMath private immutable _weightedMath;
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/managed/BaseManagedPoolFactory.sol:37]: I'd expose this

**Review by nventuro — COMMENTED** [review]
This sadly seems unavoidable :(

File: pkg/pool-weighted/contracts/ExternalWeightedMath.sol [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+/**
+ * @title A contract-wrapper for Weighted Math, Joins and Exits.
+ * @dev Use this contract as an external replacement for WeightedMath, WeightedJoinsLib and WeightedExitsLib libraries.
+ */
+contract ExternalWeightedMath {
+    /**
+     * @dev See `WeightedMath._calculateInvariant`.
+     */
+    function calcInvariant(uint256[] memory normalizedWeights, uint256[] memory balances)
```
[the remark this answers is not in the document]
Comment (jubeira) [reply pkg/pool-weighted/contracts/ExternalWeightedMath.sol:29]: Hm I went back and forth with this one; all the others start just with `calc`. But alright, let's stick to the library name.

**Review by TomAFrench — APPROVED** [review]
:megasadge:

**Pull request merged**